stdlib/atomic.mli

```diff
@@ -0,0 +1,58 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*             Stephen Dolan, University of Cambridge                     *)
+(*             Gabriel Scherer, projet Partout, INRIA Paris-Saclay        *)
+(*                                                                        *)
+(*   Copyright 2020 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** This module provides a purely sequential implementation of the
+    concurrent atomic references provided by the Multicore OCaml
+    standard library:
+
+    https://github.com/ocaml-multicore/ocaml-multicore/blob/parallel_minor_gc/stdlib/atomic.mli
+
+    This sequential implementation is provided in the interest of
+    compatibility: when people will start writing code to run on
+    Multicore, it would be nice if their use of Atomic was
+    backward-compatible with older versions of OCaml without having to
+    import additional compatibility layers. *)
+
+(** An atomic (mutable) reference to a value of type ['a]. *)
+type 'a t
+
+(** Create an atomic reference. *)
+val make : 'a -> 'a t
+
+(** Get the current value of the atomic reference. *)
+val get : 'a t -> 'a
+
+(** Set a new value for the atomic reference. *)
+val set : 'a t -> 'a -> unit
+
+(** Set a new value for the atomic reference, and return the current value. *)
+val exchange : 'a t -> 'a -> 'a
+
+(** [compare_and_set r seen v] sets the new value of [r] to [v] only
+    if its current value is physically equal to [seen] -- the
+    comparison and the set occur atomically. Returns [true] if the
+    comparison succeeded (so the set happened) and [false]
+    otherwise. *)
+val compare_and_set : 'a t -> 'a -> 'a -> bool
+
+(** [fetch_and_add r n] atomically increments the value of [r] by [n]. *)
```

Comment:
We need to document the return value.  There are two perfectly sensible results: the value of [r] before the increment, or the value after the increment.

Reply:
Yes, and initially I got it wrong, I assumed it would be the one after the increment (there is no documentation in the Multicore implementation). Returning the previous value is similar to the compare-and-swap idiom so I think it is more natural to concurrent people.

Reply:
C has fetch_and_add and add_and_fetch, with unsurprising meanings...

Reply:
Done.

Personally I find that returning the post-increment value would be more natural, but Wikipedia documents the pre-increment return: https://en.wikipedia.org/wiki/Fetch-and-add

Reply:
@maranget: nice! We could add `add_and_fetch` as well, but this requires a PR on the Multicore front as well to keep in sync. I would rather do this separately.

Reply:
As far as I can tell, C11 supports only `atomic_fetch_add`, with `fetch_and_add` semantics. I also find pre-increment a bit weird, but it is much more widespread than the reverse. Having both might be convenient, but they'd be easy to mix up when porting code.

(GCC supports both __atomic_fetch_add and __atomic_add_fetch as well as the older __sync_fetch_and_add and __sync_add_and_fetch, but none of these are portable)

Reply:
@stedolan if I understand correctly, this matters because Multicore-OCaml relies on C11 intrisics to implement the functions in the runtime system?

We could implement one of top of the other: unless I misunderstand something, `add_and_fetch r n` can always be implemented as `fetch_and_add r n + n`, and this is fairly efficient (probably no noticeable difference from just the C call).

Reply:
One thing I personally like in having both in the public API is that it clarifies the semantics of both. (And also I would naively expect `fetch_and_add` to be useful in many scenarios.) This could justify having both even if they are not both primitive.

Reply:
No, as you say they're trivial to express in terms of each other. There's no technical difficulties with any option here, just API design.

Reply:
Please, can we stick with a small, easy-to-remember API and with the primitives that are most widely used in the field?  If everyone skilled in the art of lock-free algorithms uses `fetch_and_add`, let's not add mental noise with `add_and_fetch`.

Reply:
@stedolan has completed my remark, I was referring to gcc _atomic builtins_ Let us keep it simple and follow C11 choice.